cmd/sentinel/sentinel.go

```diff
@@ -360,6 +360,8 @@ func (s *Sentinel) setDBSpecFromClusterSpec(cd *cluster.ClusterData) {
 			db.Spec.FollowConfig.StandbySettings = clusterSpec.StandbySettings
 		}
 		db.Spec.AdditionalWalSenders = *clusterSpec.AdditionalWalSenders
+		db.Spec.AdditionalSyncStandbyNames = clusterSpec.AdditionalSyncStandbyNames
```

Comment:
Don't define `db.Spec.AdditionalSyncStandbyNames` but just add the additional synchronous standbys to db.Spec.SynchronousStandby.

Reply:
@sgotti We want these additional standbys not to be taken into account here keeper.go:303

			parameters["synchronous_standby_names"] = fmt.Sprintf("%d (%s)", syncStandbyNumber, strings.Join(synchronousStandbys, ","))
		} else {
			parameters["synchronous_standby_names"] = strings.Join(synchronousStandbys, ",")
		}

Reply:
@saranhada I don't understand why. As the name states they are `AdditionalSyncStandbyNames` so they should be added to the list of "synchronous_standby_names" no?

Reply:
@saranhada If you want them only as a "fallback" standby when one or more keeper standby isn't working then that's another story and it should be handled differently since this will break the current concistency logic that uses a fake standby as this change will skip it.

As I explained [previously]( https://github.com/sorintlab/stolon/pull/410#pullrequestreview-89021195) these are 2 distinct changes. One is to set additional sync standbys (not replacement standbys) that will be counted in the number of requires sync standbys. Another is to disable sync repl when you don't have enough standbys.

Reply:
@sgotti if I add the additional standby names into db.Spec.SynchronousStandby then the provided piece of will be generated like (for instance) = "2 (stolon_fdg4gs, add_name)", and this is what I want to avoid. We want to implement additional synchronous standbys for needs of backups as it was mentioned on the schema (before). It is supposed that PgBarman will serve the required needs. Thus we add PgBarman into 2 nodes stolon cluster with enabled sync replication. In a moment the master fails the standby keeper will be promoted into new Master with sync replication still enabled. The single stolon node will still operate and the PgBarman will be used as sync standby destination.

Later when a new 2nd node of stolon is restored/created the cluster will be switched to use it. But PgBarman will serve the needs of sync backup.

Reply:
@saranhada Looks like you have two requirements:

a. You don't want that a primary blocks when there are less than `MinSynchronousStandbys` working standbys AND **you don't want to just fallback to async repl**  but instead you want to send wal records to an external standby that will become Synchronous ONLY when the above condition happen.

Note that this isn't #330 (deactivate synchronous replication for cluster consisting of 2 nodes in case the slave fails).

and

b. You want an external synchronous standby for backups with rpo = 0 but only when there are less than `MinSynchronousStandbys` working standbys.

But I'm quite sure they can be summed in only one requirement:

* Add an option to define N **Fallback** synchronous standbys that will be added only when there are less than `MinSynchronousStandbys` working standbys.

The main issue with your implementation is that you can't control all of this using the FIRST method of the `sync_standby_names` parameters because it's acting outside stolon control (switching sync stanbys without the sentinel knowing this). Let me explain:

This PR will define for the primary instance something like this: `1 (stolon_fdg4gs, add_name)`.

This means that if the standby keeper `stolon_fdg4gs` fails, the master instead of blocking will continue committing transactions since it can send data to pgBarman (or any other external pg instance defined by `add_name`).

If in a small time window a standby dies then the primary dies and the failed standby comes back before the sentinel notices this, it may happen that the standby is elected as master also if it's not in sync with the dead primary causing the loss of all the changes contained in the missed wal records.

That's the reason why we are forced to have FIRST value equal to the total number sync standbys. (BTW this check https://github.com/sorintlab/stolon/pull/410/files#diff-041abc9a078974534612e6fc3690c2bbL287we is not needed and we should just use this syntax: https://github.com/sorintlab/stolon/pull/410/files#diff-041abc9a078974534612e6fc3690c2bbL290 (not sure why I used two formats that means the same). Will open a PR to remove this confusion and document it better).

That's the reason why I was asking to just add the `AdditionalSyncStandbys` to `db.Spec.SynchronousStandbys`. But obviously this will block your primary if just one of the external sync standbys are down and you don't want this.

## Proposed implementation

So the starting point is that it must be the sentinel to control all of this. And let's change the option name to some more clear like `ExternalFallbackSyncStandbys`.

A possible solution is to handle all of this in the sentinel (like we already do adding a fakestandby when there aren't enough keeper to force the primary to block): if the sentinel detects that the good standbys are less than `MinSynchronousStandbys`  then add one or more `ExternalFallbackSyncStandbys` to reach `MinSynchronousStandbys`.

So, in a two node cluster, you'll have `MinSynchronousStandbys = 1`, if the standbys dies the sentinel will notice this, remove it from the `db.Spec.SynchronousStandbys` and add the first entry in `ExternalFallbackSyncStandbys`. In this way the primary will unblock. The above problem won't happen because the external standby will be added by the sentinel only if it detects that the internal standbys is dead. If the dead standby comes back it'll be readded to the list of sync stanbys and the external stanby will be removed.

Of course the switch isn't instantaneus as if it's managed by the postgres instance but will permit the cluster to unblock after some seconds/minutes.